sync/metrics.go

```diff
@@ -11,40 +12,140 @@ import (
 		return nil, err
 	}
 
-	m := &metrics{
-		totalSyncedGauge: totalSynced,
+	trustedPeersOutOfSync, err := meter.Int64Counter(
+		"hdr_sync_trust_peers_out_of_sync",
+		metric.WithDescription("trusted peers out of sync and gave outdated header"),
+	)
+	if err != nil {
+		return nil, err
 	}
 
-	callback := func(ctx context.Context, observer metric.Observer) error {
-		observer.ObserveFloat64(totalSynced, float64(m.totalSynced.Load()))
-		return nil
+	unrecentHeader, err := meter.Int64Counter(
+		"hdr_sync_unrecent_header",
+		metric.WithDescription("tracks every time Syncer returns an unrecent header"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	subjectiveInit, err := meter.Int64Counter(
+		"hdr_sync_subjective_init",
+		metric.WithDescription(
+			"tracks how many times is the node initialized ",
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	subjectiveHead, err := meter.Int64ObservableGauge(
+		"hdr_sync_subjective_head",
+		metric.WithDescription("subjective head height"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	blockTime, err := meter.Float64Histogram(
+		"hdr_sync_actual_blockTime_ts",
+		metric.WithDescription("duration between creation of 2 blocks"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &metrics{
+		syncLoopStarted:       syncLoopStarted,
+		trustedPeersOutOfSync: trustedPeersOutOfSync,
+		unrecentHeader:        unrecentHeader,
+		subjectiveInit:        subjectiveInit,
+		blockTime:             blockTime,
+		subjectiveHeadInst:    subjectiveHead,
 	}
-	_, err = meter.RegisterCallback(callback, totalSynced)
+
+	m.subjectiveHeadReg, err = meter.RegisterCallback(m.newHead, m.subjectiveHeadInst)
 	if err != nil {
 		return nil, err
 	}
 
 	return m, nil
 }
 
-// recordTotalSynced records the total amount of synced headers.
-func (m *metrics) recordTotalSynced(totalSynced int) {
+func (m *metrics) newHead(_ context.Context, obs metric.Observer) error {
+	obs.ObserveInt64(m.subjectiveHeadInst, m.subjectiveHead.Load())
+	return nil
+}
+
+func (m *metrics) syncingStarted(ctx context.Context) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.syncLoopStarted.Add(ctx, 1)
+	})
+}
+
+func (m *metrics) laggingNetworkHead(ctx context.Context) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.unrecentHeader.Add(ctx, 1)
+	})
+}
+
+func (m *metrics) trustedPeersOutOufSync(ctx context.Context) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.trustedPeersOutOfSync.Add(ctx, 1)
+	})
+}
+
+func (m *metrics) subjectiveInitialization(ctx context.Context) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.subjectiveInit.Add(ctx, 1)
+	})
+}
+
+func (m *metrics) newSubjectiveHead(ctx context.Context, height uint64, timestamp time.Time) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.subjectiveHead.Store(int64(height))
+
+		if !m.prevHeader.IsZero() {
+			m.blockTime.Record(ctx, timestamp.Sub(m.prevHeader).Seconds())
+		}
+	})
+}
+
+func (m *metrics) observe(ctx context.Context, observeFn func(context.Context)) {
```

Comment:
it checks whether `m` is initialized or just created as an empty struct, since observeFunc can hold fields of the struct that can cause panic:

```
	if m == nil {
		return
	}
```